File: src/devtwin/adapters/ruby.py
"""Ruby ecosystem adapter."""

from __future__ import annotations

from pathlib import Path

from devtwin.adapters.base import EcosystemAdapter, extract_version
from devtwin.core.models import DependencyInfo, HealthIssue, Presence, RuntimeInfo, Severity
from devtwin.core.runner import run_command, which


class RubyAdapter(EcosystemAdapter):
    ecosystem = "ruby"

    def detect(self, root: Path) -> bool:
        if (root / "Gemfile").exists():
            return True
        return bool(list(root.glob("*.gemspec")))

    def _required_version(self, root: Path) -> str | None:
        f = root / ".ruby-version"
        if not f.exists():
            return None
        try:
            return f.read_text(encoding="utf-8", errors="ignore").strip() or None
        except OSError:
            return None

    def inspect_runtime(self, root: Path) -> list[RuntimeInfo]:
        required = self._required_version(root)
        ruby_path = which("ruby")
        if ruby_path is None:
            return [
                RuntimeInfo(
                    name="ruby",
                    presence=Presence.NOT_INSTALLED,
                    required_version=required,
                    source=".ruby-version" if required else None,
                )
            ]
        result = run_command([ruby_path, "--version"], timeout=5)
        version = extract_version(result.stdout) if result.available else None
        return [
            RuntimeInfo(
                name="ruby",
                presence=Presence.INSTALLED if version else Presence.UNKNOWN,
                installed_version=version,
                required_version=required,
                path=ruby_path,
                source=".ruby-version" if required else None,
            )
        ]

    def inspect_dependencies(self, root: Path) -> DependencyInfo | None:
        if not (root / "Gemfile").exists():
            return None
        lockfile = root / "Gemfile.lock"
        return DependencyInfo(
            ecosystem="ruby",
            manager="bundler",
            lockfile="Gemfile.lock" if lockfile.exists() else None,
            lockfile_present=lockfile.exists(),
            manifest_present=True,
        )

    def inspect_tests(self, root: Path) -> list[str]:
        if (root / "spec").exists():
            return ["bundle exec rspec"]
        if (root / "Rakefile").exists():
            return ["bundle exec rake test"]
        return []

    def health_checks(self, root: Path, runtimes: list[RuntimeInfo]) -> list[HealthIssue]:
        issues = []
        for rt in runtimes:
            if rt.name == "ruby" and rt.presence == Presence.NOT_INSTALLED:
                issues.append(
                    HealthIssue(
                        severity=Severity.CRITICAL,
                        code="ruby.not_installed",
                        title="Ruby not found",
                        message="No `ruby` executable was found on PATH.",
                        evidence=["which ruby -> not found"],
                        recommendation="Install Ruby via rbenv, rvm, or https://www.ruby-lang.org/en/downloads.",
                    )
                )
        return issues

File: mcphub/devtwin/src/adapters/ruby.ts
/** Ruby ecosystem adapter. */

import { join } from 'node:path';

import {
  Presence,
  Severity,
  type DependencyInfo,
  type HealthIssue,
  type RuntimeInfo,
} from '../core/models.js';
import { runCommand, which } from '../core/runner.js';
import { globAny, pathExists, readTextFile } from '../system/filesystem.js';
import { EcosystemAdapter, extractVersion } from './base.js';

export class RubyAdapter extends EcosystemAdapter {
  readonly ecosystem = 'ruby';

  detect(root: string): boolean {
    if (pathExists(join(root, 'Gemfile'))) {
      return true;
    }
    return globAny(root, ['*.gemspec'], 1).length > 0;
  }

  private requiredVersion(root: string): string | null {
    const file = join(root, '.ruby-version');
    if (!pathExists(file)) {
      return null;
    }
    // readTextFile already degrades to null on any read error (the Python
    // equivalent swallows OSError here).
    const text = readTextFile(file);
    if (text === null) {
      return null;
    }
    return text.trim() || null;
  }

  async inspectRuntime(root: string): Promise<RuntimeInfo[]> {
    const required = this.requiredVersion(root);
    const rubyPath = which('ruby');
    if (rubyPath === null) {
      return [
        {
          name: 'ruby',
          presence: Presence.NOT_INSTALLED,
          installed_version: null,
          required_version: required,
          path: null,
          source: required ? '.ruby-version' : null,
        },
      ];
    }
    const result = await runCommand([rubyPath, '--version'], { timeout: 5 });
    const version = result.available ? extractVersion(result.stdout) : null;
    return [
      {
        name: 'ruby',
        presence: version ? Presence.INSTALLED : Presence.UNKNOWN,
        installed_version: version,
        required_version: required,
        path: rubyPath,
        source: required ? '.ruby-version' : null,
      },
    ];
  }

  async inspectDependencies(root: string): Promise<DependencyInfo | null> {
    if (!pathExists(join(root, 'Gemfile'))) {
      return null;
    }
    const lockfilePresent = pathExists(join(root, 'Gemfile.lock'));
    return {
      ecosystem: 'ruby',
      manager: 'bundler',
      lockfile: lockfilePresent ? 'Gemfile.lock' : null,
      lockfile_present: lockfilePresent,
      manifest_present: true,
      installed: Presence.UNKNOWN,
      dependency_count: null,
      dev_dependency_count: null,
      notes: [],
    };
  }

  async inspectTests(root: string): Promise<string[]> {
    if (pathExists(join(root, 'spec'))) {
      return ['bundle exec rspec'];
    }
    if (pathExists(join(root, 'Rakefile'))) {
      return ['bundle exec rake test'];
    }
    return [];
  }

  async healthChecks(_root: string, runtimes: RuntimeInfo[]): Promise<HealthIssue[]> {
    const issues: HealthIssue[] = [];
    for (const rt of runtimes) {
      if (rt.name === 'ruby' && rt.presence === Presence.NOT_INSTALLED) {
        issues.push({
          severity: Severity.CRITICAL,
          code: 'ruby.not_installed',
          title: 'Ruby not found',
          message: 'No `ruby` executable was found on PATH.',
          evidence: ['which ruby -> not found'],
          recommendation:
            'Install Ruby via rbenv, rvm, or https://www.ruby-lang.org/en/downloads.',
          confidence: null,
        });
      }
    }
    return issues;
  }
}
